rstar/src/primitives/geom_with_data.rs

```diff
@@ -0,0 +1,136 @@
+use crate::envelope::Envelope;
+use crate::object::PointDistance;
+use crate::{object::RTreeObject, point::Point};
+
+/// An [RTreeObject] with a geometry and some associated data that can be inserted into an r-tree.
+///
+/// Often, adding metadata (like a database ID) to a geometry is required before adding it
+/// into an r-tree. This struct removes some of the boilerplate required to do so.
+///
+/// **Note:** while the container itself implements [RTreeObject], you will have to go through its
+/// [`geom`](GeomWithData::geom) method in order to access geometry-specific methods.
+///
+/// # Example
+/// ```
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+///
+/// type RestaurantLocation = GeomWithData<[f64; 2], &'static str>;
+///
+/// let mut restaurants = RTree::new();
+/// restaurants.insert(RestaurantLocation::new([0.3, 0.2], "Pete's Pizza Place"));
+/// restaurants.insert(RestaurantLocation::new([-0.8, 0.0], "The Great Steak"));
+/// restaurants.insert(RestaurantLocation::new([0.2, -0.2], "Fishy Fortune"));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest restaurant!
+/// let place = restaurants.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.data);
+/// println!("It's really close, only {} miles", place.distance_2(&my_location));
+/// ```
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct GeomWithData<R, T> {
```

Comment:
```suggestion
pub struct GeomWithData<R: RTreeObject, T> {
```

One more minor thing - folks might get a *slightly* clearer error message if they mix up these params if we also constrain the struct definition, rather than only constraining the impls.

I can't think of a reason we'd want someone to be able to build a GeomWithData where R was *not* an RTreeObject, right?

Reply:
Something like 335a029? I _did_ wonder why that hadn't been done for `PointWithData`…

Reply:
Yep, that's exactly what I had in mind!